tephra-core/src/main/java/co/cask/tephra/distributed/ElasticPool.java

```diff
@@ -107,22 +112,23 @@ public ElasticPool(int sizeLimit) {
    *
    * @return an element
    */
-  public T obtain() throws E {
-    T element = getOrCreate();
-    while (true) {
-      if (element != null) {
-        return element;
-      }
-      synchronized (this) {
-        try {
-          this.wait();
-        } catch (InterruptedException e) {
-          LOG.warn("Wait interrupted. Don't know what to do. Ignoring.");
-          continue;
-        }
-        // got notified - try to get element
-        element = getOrCreate();
+  public T obtain() throws E, TimeoutException, InterruptedException {
+    if (!semaphore.tryAcquire(1, obtainClientTimeoutMs, TimeUnit.MILLISECONDS)) {
+      throw new TimeoutException(String.format("Failed to obtain client within %s milliseconds.",
+                                               obtainClientTimeoutMs));
+    }
+    try {
+      T client = elements.poll();
+      // a client was available, all good. otherwise, create one
+      if (client != null) {
+        return client;
       }
+      return create();
+    } catch (Exception e) {
```

Comment:
By catching `Exception`, we still catch runtime exceptions, we just omit `Error`s.